apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,219 @@
+import {
+  Injectable,
+  NestInterceptor,
+  ExecutionContext,
+  CallHandler,
+  Logger,
+  ConflictException,
+  HttpException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { CacheService } from '@novu/application-generic';
+import { Observable, of, throwError } from 'rxjs';
+import { catchError, map } from 'rxjs/operators';
+import { createHash } from 'crypto';
+
+const LOG_CONTEXT = 'IdempotencyInterceptor';
+const IDEMPOTENCY_CACHE_TTL = 60 * 60 * 6; //6h
+const IDEMPOTENCY_PROGRESS_TTL = 60 * 5; //5min
+
+const HEADER_KEYS = {
+  IDEMPOTENCY: 'idempotency-key',
+  RETRY_AFTER: 'retry-after',
+  IDEMPOTENCY_CONFLICT: 'x-idempotency-conflict',
+};
+
+enum ReqStatusEnum {
+  PROGRESS = 'in-progress',
+  SUCCESS = 'success',
+  ERROR = 'error',
+}
+
+@Injectable()
+export class IdempotencyInterceptor implements NestInterceptor {
+  constructor(private readonly cacheService: CacheService) {}
+
+  async intercept(context: ExecutionContext, next: CallHandler): Promise<Observable<any>> {
+    const request = context.switchToHttp().getRequest();
+    const idempotencyKey = this.getIdempotencyKey(context);
+    const isEnabled = process.env.API_IDEMPOTENCY_ENABLED == 'true';
+    if (!isEnabled || !idempotencyKey || !['post', 'patch'].includes(request.method.toLowerCase())) {
+      return next.handle();
+    }
+    const cacheKey = this.getCacheKey(context);
+
+    try {
+      const bodyHash = this.hashRequestBody(request.body);
+      //if 1st time we are seeing the request, marks the request as in-progress if not, does nothing
+      const isNewReq = await this.setCache(
+        cacheKey,
+        { status: ReqStatusEnum.PROGRESS, bodyHash },
+        IDEMPOTENCY_PROGRESS_TTL,
+        true
+      );
+      // Check if the idempotency key is in the cache
+      if (isNewReq) {
+        return await this.handleNewRequest(context, next, bodyHash);
+      } else {
+        return await this.handlerDuplicateRequest(context, bodyHash);
+      }
+    } catch (err) {
+      Logger.warn(
+        `An error occurred while making idempotency check, proceeding without idempotency. key:${idempotencyKey}. error: ${err.message}`,
+        LOG_CONTEXT
+      );
+    }
+
+    //something unexpected happened, both cached response and handler did not execute as expected,so proceed w/o idempotency
+    return next.handle();
+  }
+
+  private getIdempotencyKey(context: ExecutionContext): string | undefined {
+    const request = context.switchToHttp().getRequest();
+
+    return request.headers[HEADER_KEYS.IDEMPOTENCY];
+  }
+
+  private getCacheKey(context: ExecutionContext): string {
+    const request = context.switchToHttp().getRequest();
+    const { organizationId } = request.user || {};
```

Comment:
@mahendraHegde have you tested this with using an ApiKey request? This results in undefined when making a request with an API key.